### PR TITLE
FEATURE: Preserve cursor in editor upload

### PR DESCRIPTION
**Pull request by caugner**
See: https://meta.discourse.org/t/cursor-is-overriden-when-upload-finishes/54892?u=claas

### CODE REVIEW

File: app/assets/javascripts/discourse/components/d-editor.js.es6 [hunk 1]
```diff
@@ -525,11 +525,43 @@ export default Ember.Component.extend({
...
+    const selectionStart = textarea.selectionStart;
+    const selectionEnd = textarea.selectionEnd;
+    const needleEnd = needleStart + oldValue.length;
+    const replacementEnd = needleStart + newVal.length;
+
+    // Replace value (side effect: cursor at end).
+    this.set('value', val.replace(oldVal, newVal));
+
+    // Determine cursor/selection.
+    const newSelectionStart, newSelectionEnd;
```
Comment (caugner) [inline app/assets/javascripts/discourse/components/d-editor.js.es6]: Not a good idea, will fix.

File: app/assets/javascripts/discourse/components/d-editor.js.es6 [hunk 2]
```diff
@@ -525,11 +525,43 @@ export default Ember.Component.extend({
...
+      // Nothing to replace.
+      return;
     }
+
+    const textarea = this.$('textarea.d-editor-input')[0];
+
+    // Remember cursor/selection.
+    const selectionStart = textarea.selectionStart;
+    const selectionEnd = textarea.selectionEnd;
+    const needleEnd = needleStart + oldValue.length;
```
Comment (caugner) [inline app/assets/javascripts/discourse/components/d-editor.js.es6]: Note to myself: Always test again after final refactoring.

**Comment by caugner** [comment]
There's a [test which checks the current behavior](https://github.com/discourse/discourse/blob/5dbd6a304bed5400be481d71061d3e3ebb4d6785/test/javascripts/components/d-editor-test.js.es6#L763-L776):

```
Module Failed: component:d-editor
  Test Failed: replace-text event
    Assertion Failed:
      Expected: 10, Actual: 0    
    Assertion Failed:
      Expected: 10, Actual: 0
```

I'll change the test expectation and add some more tests.

File: app/assets/javascripts/discourse/components/d-editor.js.es6 [hunk 3]
```diff
@@ -525,11 +525,43 @@ export default Ember.Component.extend({
...
+
+    if (needleStart === -1) {
+      // Nothing to replace.
+      return;
     }
+
+    const textarea = this.$('textarea.d-editor-input')[0];
+
+    // Remember cursor/selection.
+    const selectionStart = textarea.selectionStart;
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/components/d-editor.js.es6]: I'll prefer that we move this chunk into 

https://github.com/discourse/discourse/blob/master/app/assets/javascripts/discourse/lib/utilities.js.es6

**Pull request merged**